### ISSUE TITLE
Zuul and X-Forwarded-For

### ISSUE THREAD

**Issue by vsafin-copart** [spring-cloud/spring-cloud-netflix#352]
Should Zuul support X-Forwarded-For out of the box? Just like it sets X-Forwarded-Host

**Comment by dsyer** [comment]
Sounds like a reasonable idea (obviously checking first if the header is already set and not setting it in that case).

**Comment by timtebeek** [comment]
At present it seems the X-Forwarded-For headers that are passed into a typical `@EnableZuulProxy` `@SpringBootApplication` are stripped from the requests: https://gist.github.com/timtebeek/7aa8f6fc1b8c014c6e4a
(using `spring-cloud-starter-zuul:Angel-SR3`)

>`curl -v -H 'X-Real-IP: 3.4.5.6' -H 'X-Forwarded-For: 1.2.3.4' http://apache.proxied.zuul.instance/manage/trace`

This makes it impossible for me to access the value in any downstream backend services, as needed for some auth-logic.

What changes would I have to make to at least retain the `X-Forwarded-For` headers that are passed in?

**Comment by dsyer** [comment]
Probably some additional logic in `ProxyRequestHelper` is all that is needed.

**Comment by timtebeek** [comment]
For now I've setup a minimal ZuulFilter:
```java
	public Object run() {
		RequestContext ctx = RequestContext.getCurrentContext();
		// Rely on HttpServletRequest to retrieve the correct remote address from upstream X-Forwarded-For header
		HttpServletRequest request = ctx.getRequest();
		String remoteAddr = request.getRemoteAddr();
		// Pass remote address downstream by setting X-Forwarded for header again on Zuul request
		log.debug("X-Forwarded-For set to: {}", remoteAddr);
		ctx.getZuulRequestHeaders().put(X_FORWARDED_FOR, remoteAddr);
		return null;
	}
```
That works, but it seems such a common case to me that might warrant better handling. :)

**Comment by dsyer** [comment]
Were you going to send a pull request then?

**Comment by timtebeek** [comment]
Hadn't thought about it yet, as I thought the handling might need to be extended a bit to conform to the behavior seen in other proxies, such as appending a new value for `X-Forwarded-For` rather than merely sending the old value. Didn't yet think I'd get that right the first time in a pull request.

**Comment by dsyer** [comment]
Fair enough. I think the default should be what you did (replace the existing one), and we can add a boolean flag option to append as well.

**Comment by timtebeek** [comment]
How does this tie into [Spring Boot 1.3 X-Forwarded header support](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-1.3-Release-Notes#x-forwarded-for-header-support)?

**Comment by dsyer** [comment]
That's related to the backend, so I don't think there is any tie in really.

**Comment by dsyer** [comment]
Our Zuul `PreDecorationFilter` already sends `x-forwarded-*` headers by default. I don't think there's anything else to do here.

**Comment by timtebeek** [comment]
@dsyer It's true that some `X-Forwarded-` headers are now set, but `X-Forwarded-For` appears not to be set still: https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/pre/PreDecorationFilter.java

**Comment by dsyer** [comment]
OK, I guess we are still waiting for a pull request then. Maybe we should append rather than replace though, by default?

**Comment by timtebeek** [comment]
@dsyer Please have a look at #952

**Comment by dsyer** [comment]
Fixed via #952
